src/crystal/system/win32/file.cr

```diff
@@ -26,11 +26,85 @@ module Crystal::System::File
   end
 
   def self.open(filename : String, flags : Int32, perm : ::File::Permissions) : {LibC::Int, Errno}
-    flags |= LibC::O_BINARY | LibC::O_NOINHERIT
+    access, disposition, attributes, share = self.posix_to_open_opts flags, perm
 
-    fd = LibC._wopen(System.to_wstr(filename), flags, perm)
+    handle = LibC.CreateFileW(
+      System.to_wstr(filename),
+      access,
+      share, # UNIX semantics
+      nil,
+      disposition,
+      attributes,
+      LibC::HANDLE.null
+    )
+
+    if handle == LibC::INVALID_HANDLE_VALUE
+      # Map ERROR_FILE_EXISTS to Errno::EEXIST to avoid changing semantics of other systems
+      if WinError.value.error_file_exists?
+        Errno.value = Errno::EEXIST
+      end
+
+      return {-1, Errno.value}
+    end
+
+    fd = LibC._open_osfhandle handle, flags
+
+    if fd == -1
+      return {-1, Errno.value}
+    end
+
+    if flags & LibC::O_BINARY > 0
+      LibC._setmode fd, LibC::O_BINARY
+    elsif flags & LibC::O_TEXT > 0
+      LibC._setmode fd, LibC::O_TEXT
+    end
+
+    {fd, Errno::NONE}
+  end
+
+  private def self.posix_to_open_opts(flags : Int32, perm : ::File::Permissions)
+    disposition = 0
+
+    access = if flags & LibC::O_WRONLY > 0
+               LibC::GENERIC_WRITE
+             elsif flags & LibC::O_RDWR > 0
+               LibC::GENERIC_READ | LibC::GENERIC_WRITE
+             else
+               LibC::GENERIC_READ
+             end
+
+    if flags & LibC::O_APPEND > 0
+      access |= LibC::FILE_APPEND_DATA
+    end
+
+    case flags & (LibC::O_CREAT | LibC::O_EXCL | LibC::O_TRUNC)
+    when 0, LibC::O_EXCL                                                            then disposition = LibC::OPEN_EXISTING
+    when LibC::O_CREAT                                                              then disposition = LibC::OPEN_ALWAYS
+    when LibC::O_CREAT | LibC::O_EXCL, LibC::O_CREAT | LibC::O_TRUNC | LibC::O_EXCL then disposition = LibC::CREATE_NEW
+    when LibC::O_TRUNC, LibC::O_TRUNC | LibC::O_EXCL                                then disposition = LibC::TRUNCATE_EXISTING
+    when LibC::O_CREAT | LibC::O_TRUNC                                              then disposition = LibC::CREATE_ALWAYS
+    end
```

Comment:
I find it confusing to decipher which input flag leads to which output flag.
This is what I came up with trying to unfiddle this. I don't think it's necessarily more readable. Not sure about performance either. Just posting it as a discussion entry.
```suggestion
   if flags & LibC::O_TRUNC > 0
     if flags & LibC::O_CREAT > 0
       disposition = LibC::CREATE_ALWAYS
     else
       disposition = LibC::TRUNCATE_EXISTING
     end
   elsif flags & LibC::O_CREAT > 0
     if flags & LibC::O_EXCL > 0
       distribution = LibC::CREATE_NEW
     else
       distribution = LibC::OPEN_ALWAYS
     end
   else
     distribution = LibC::OPEN_EXISTING
   end
```